### ISSUE TITLE
☂️ linter support for views 

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#3625]
A meta-issue to discuss and track work on linter support for views.

(**TODO**: add a link to the spec when one is finalized.  In the meantime, the discussion in https://github.com/dart-lang/language/issues/2360 is a good place to get some context.)

**Comment by IvanDembicki** [comment]
avoid_logic_in_view_classes

**Comment by bwilkerson** [comment]
@IvanDembicki Could you provide more information about what the proposed lint would actually do (either here or in a separate issue, the latter being my personal preference).

**Comment by IvanDembicki** [comment]
> 
[bwilkerson](https://github.com/bwilkerson) #3633

**Comment by pq** [comment]
@eernstg: in the spec you write:

> we make no attempt to hide the representation type during the exhaustiveness analysis. The usage of such patterns is very similar to a cast into the extension type in the sense that it provides a reference of type V to an object without running any constructors of V. We will rely on lints in order to inform developers about such situations, similarly to the treatment of expressions like e as V.

Could you provide a bit of detail on the lints you had imagined?

**Comment by eernstg** [comment]
@pq wrote:
> Could you provide a bit of detail on the lints you had imagined?

Sure! Ideally, they would flag any construct that performs a dynamic type test on an expression `e` and yields an expression whose type is an extension type which is not the static type of `e` nor a supertype thereof. That is, anything which is a cast or a promotion "into" an extension type.

We need to consider types where extension types occur as subterms as well. For instance, if `E` is an extension type then `List<E>` is relevant as well, and so is `E Function(Map<E, String>)`, etc.

So let's spell out this concept in more detail:

We say that a type `T` _contains an extension type_ `E` in the case where `E` is an extension type and `T` is a term that contains `E` as a subterm. We say that a type `T` _contains a covariant extension type_ in the case where `T` contains an extension type `E` that occurs covariantly, and similarly for _contains a contravariant extension type_ and _contains an invariant extension type_.

We say that the transition from a source type `S` to a destination type `T` _introduces an extension type_ `E` if:

- `T` contains a covariant or invariant extension type `E`, and `S` is not a subtype of `T` nor a supertype of `T`. *For example, a transition from `int` to `E`, where `E` is an extension type with no subtype relationship to `int`.*
- `T <: S`, `T` contains a covariant or invariant extension type `E`, and the proof of `T <: S` contains a subproof of the form `E <: U` where `U != E`. *For example, a transition from `Iterable<int>` to `List<E>` where `E <: int`.; or a transition from `Object? Function(int)` to `E Function(int)`.*
- `T <: S`, `T` contains a covariant or invariant extension type `E`, and the proof of `T <: S` contains a subproof where the structure of the types eliminates the type `E`. *For instance, a transition from `Object` to `List<E>`; or a transition from `Function` to `void Function({String Function(E) callback})`.*

*Note that an upcast never introduces an extension type, even in cases like the transition from `void Function(Object?)` to `void Function(E)` where `E` is an extension type. The point is that no object will be treated as an instance of `E` which wouldn't have been treated as such before the transition. For example, in the simple case `((Object? o) => e) as void Function(E)`, the body `e` still treats the argument as having the type `Object?`. Each call site must new provide an argument with static type `E`, but it's a separate issue how we managed to create an expression with that type.*

With that, we can list cases that are candidates to be linted:

- `e as T`, where the transition from the static type of `e` to `T` introduces an extension type.
- `e is T`, where the transition from the static type of `e` to `T` introduces an extension type.
- `on T` in a try/catch statement, where `T` contains an extension type.
- Matching against an object pattern whose type `T` contains a covariant or invariant extension type (in any context where a pattern can occur), in a situation where the transition from the matched value type to `T` introduces an extension type.
- Implicit casts from `dynamic` to a type that contains an extension type.

Finally, the lint message should be omitted in the case where the extension type which was introduced has a direct or indirect superinterface which is a non-extension type. *For example, `extension type E(T t) implements S {...}`, where `S` is a non-extension type.*

*You could say that `implements S` is a way for `E` to announce to the world that _every_ object whose dynamic type is the representation type of `E` or a subtype thereof is an appropriate representation object for `E`. In contrast, an extension type that does not implement any other non-extension types than the default ones (top types for some extension types, `Object` for others) provide a soft kind of encapsulation in that we're not supposed to cast or promote or match our way into those extension types, and that's exactly what the lint is supposed to detect.*

**Comment by pq** [comment]
Regarding enforcing (CamelCase) naming, I'd love not to add another lint but would love to get people's 2 cents.  How do folks feel about extending [`camel_case_extensions`](https://dart.dev/tools/linter-rules/camel_case_extensions) or [`camel_case_types`](https://dart.dev/tools/linter-rules/camel_case_types) to support extension types?

/cc @munificent @natebosch @eernstg @lrhn @bwilkerson @srawlins @jakemac53

**Comment by parlough** [comment]
I see no reason to not add it to an existing lint, especially if the lint is updated before extension types land in stable.

It seems it _(and `camel_case_extensions`)_ should be handled by `camel_case_types` to me.

**Comment by pq** [comment]
Thanks @parlough! I agree we may have been too conservative w/ `camel_case_extensions` (@bwilkerson may recall any added context) and maybe we should fold them all together down the road but yeah, that's a bit tricky...

**Comment by pq** [comment]
Now, if you had to **choose** _which_ lint to extend?

**Comment by parlough** [comment]
> Now, if you had to choose which lint to extend?

Unless that potential added context changes something, I would say `camel_case_types`. That's closer to an all encompassing term for the future, and "extension types" has "types" in it as well :)

**Comment by bwilkerson** [comment]
An extension type defined a type; an extension does not.

I don't remember why we created `camel_case_extensions`, but using `camel_case_types` to enforce a declaration that doesn't introduce a type might have been the motivating factor.

**Comment by pq** [comment]
Regarding the possible constness of the primary constructor ([spec](https://github.com/dart-lang/language/blob/main/accepted/future-releases/extension-types/feature-specification.md#is-the-primary-constructor-constant)), I wonder if `const`-enthusiasts would value a new lint to nudge authors to declare them so.

That is:


**BAD**
```dart
extension type E(int i) { }
```

**GOOD**
```dart
extension type const E(int i) { }
```

(See also the conversation about immutability and extension types as it pertains to linting in #4720 and #4722.)

/cc @eernstg @goderbauer @a14n @bwilkerson

**Comment by eernstg** [comment]
As mentioned above, a primary constructor of an extension type can always be constant, because it has no initializer list, and the extension type has no non-final instance variables.

However, this nudge may be considered spammy if the advice is given for every extension type declaration, especially in cases where the static type of the parameter does not have any constant constructors (we generally can't know, except for a `final` type with no reopen there could be a subtype which has a constant constructor).

So it's probably good to have the lint (and have it flag all extension type constructors which can be constant, not just the primary one), but it is probably also useful to make sure that it is a separate lint for extension types, not an extension of the scope of the corresponding lint for classes.

**Comment by srawlins** [comment]
A few more lint rules that need changes:

* `avoid_types_as_parameter_types` is not reporting extension types (`extension type ET(int it) {} void f(ET) {}`).
* `cancel_subscriptions` seems not to report for an extension type.
* `collection_methods_unrelated_type` seems to over-report, when a method is redeclared. E.g.

  ```dart
  extension type ET(List<num> it) implements Iterable<num> {
    int contains(String i) => 7;
  }
  void f(ET et) {
    et.contains('1');
  }
  ```

  This should _not_ report on the call to `contains`, as it has been redeclared.

**Comment by pq** [comment]
Fantastic. Thanks @srawlins!

**Comment by srawlins** [comment]
A few more lint rules that need changes:

* `avoid_slow_async_io`

  This code should have lint reported, but does not.

  ```dart
  import 'dart:io';
  extension type ET(Directory dir) implements FileSystemEntity {}
  void f(ET dir) async {
    await dir.exists();
  }
  ```

* `unrelated_type_equality_checks`

  This code should not have lint reported, but does:

  ```dart
  extension type ET(int it) implements num {}
  void m(int? a1, ET a2) {
    var b1 = a1 == a2;
    var b2 = a2 == a1;
  }
  ```

* `unsafe_html`

  This code should have reported lint, but does not:

  ```dart
  import 'dart:html';
  extension type ET(HeadingElement it) implements Element {}
  void f(ET heading) {
    heading.setInnerHtml('<script>');
  }
  ```

**Comment by jacob314** [comment]
These are really interesting edge cases. It has taken me a minute to wrap my head around them. The `dart:io` one seems to be a generalization of a general issue where the lint doesn't consistently trigger for subclasses of File only the precise file class.
> * `avoid_slow_async_io`
>   This code should have lint reported, but does not.
>   ```dart
>   import 'dart:io';
>   extension type ET(Directory dir) implements FileSystemEntity {}
>   void f(ET dir) async {
>     await dir.exists();
>   }
>   ```
Turns out this one is already fragile with respect to regular Dart classes. Probably a P3 bug to make it less fragile in general.
```dart
class DummyFile implements File {
  @override
  dynamic noSuchMethod(Invocation invocation) => super.noSuchMethod(invocation);
}
void main() async {
  DummyFile file = DummyFile();
  await file.exists();
```
does not trigger the lint but
```dart
File file = DummyFile();
await file.exists();
```
So the existing behavior sometimes triggers on subclasses of File but sometimes doesn't.

> * `unrelated_type_equality_checks`
Here there is an interesting question of whether the lint should or should not trigger if the "implements num" clause was missing. With "implements num" the lint shouldn't fire but without it I think this is like the similar case that can occur with unrelated interfaces. Here the lint correctly triggers even though the types really are related.
```
class A {}

class B {}

class C implements A, B {}

void main() async {
  A a = C();
  B b = C();
  if (a == b) {
    print("equal");
  }
}
``` 
>   This code should not have lint reported, but does:
>   ```dart
>   extension type ET(int it) implements num {}
>   void m(int? a1, ET a2) {
>     var b1 = a1 == a2;
>     var b2 = a2 == a1;
>   }
>   ```
> * `unsafe_html`
>   This code should have reported lint, but does not:
>   ```dart
>   import 'dart:html';
>   extension type ET(HeadingElement it) implements Element {}
>   void f(ET heading) {
>     heading.setInnerHtml('<script>');
>   }
>   ```
This lint has an existing similar issue for implementing classes. Arguably the lint should still fire for this case of a class implementing HeadingElement but it doesn't.
Example:
```dart
import 'dart:html';

class ET implements HeadingElement {
  @override
  dynamic noSuchMethod(Invocation invocation) => super.noSuchMethod(invocation);
}

void f(ET heading) {
  heading.setInnerHtml('<script>');
}
```

**Comment by srawlins** [comment]
A few more lint rules that need changes:

* [`close_sinks`](https://dart.dev/tools/linter-rules/close_sinks) - This code should report a lint but does not:

  ```dart
  import 'dart:io';
  extension type E(IOSink it) {}
  E get e => E(File('').openWrite());
  class C {
    // ignore: unused_field
    late E _sink;
    void init() {
      _sink = e;
    }
  }
  ```
* [`avoid_setters_without_getters`](https://dart.dev/tools/linter-rules/avoid_setters_without_getters) - I opened https://github.com/dart-lang/linter/issues/4757 which goes beyond what was implemented in https://github.com/dart-lang/linter/issues/4710
* [`avoid_catching_errors`](https://dart.dev/tools/linter-rules/avoid_catching_errors) - This code should report a lint but does not:

  ```
  extension type E(Error it) {}
  void f() {
    try {} on E {}
  }
  ```
* [`only_throw_errors`](https://dart.dev/tools/linter-rules/only_throw_errors) - I _think_ think should not report a lint, but it does:

  ```dart
  extension type E(ArgumentError it) {}
  void f(E e) {
    throw e;
  }
  ```

  I guess technically an E is not a subtype of Error or Exception, but it seems like it should be fine to throw such things?

**Comment by bwilkerson** [comment]
This is interesting.

The `avoid_catching_errors` and `only_throw_errors` seem inconsistent. In neither case is `E` a subtype of `Error`. I think the question here is whether the lint should use the static type or the "runtime" type. We generally stick to the static type because it's knowable, but it sounds like in `avoid_catching_errors` case you're arguing in favor of using the runtime type. Does/should it flag
```
try {} on Object {}
```
which could also be catching `Error`s? Maybe extension types are special because we know more about the runtime type?

The question about `close_sinks` is the same. The proposal appears to be to use the runtime type rather than the static type.

If we want to go that direction, I'd want to understand better when we should use the runtime type rather than the static type so that we can be more consistent about where that technique is applied.

**Comment by eernstg** [comment]
Here is one way to think about the representation type of an extension type that might be useful when deciding on things like "should `only_throw_errors` flag `throw e` when the static type of `e` is a given extension type?":

_Check whether the associated semantics includes a dynamic type test or an upcast to a top type or `Object`._

The point is that if the semantics of a construct discards the extension type then there is no way we can restore it (there is no evidence whatsoever for the extension type at run time). In that case it makes sense to treat the expression as if it had had the representation type as its static type (and it should be the recursively erased type that we sometimes call the 'ultimate' representation type). We're trying to take certain scenarios into account, and they may depend on the run-time type, and the ultimate representation type is our best compile-time approximation of that run-time type, but the scenarios certainly won't depend on the extension type.

For example, evaluation of `throw e` is linked to an upcast to `Object` and/or a dynamic type test, because the value of `e` will be handled in code where its static type is `Object`, or it will be subject to a dynamic type test (`on T catch ...`). A future which is completed with an error will also give the thrown object the static type `Object`.

---

Now what if we have the property mentioned above (i.e., the extension type is unavoidably discarded), but the most _useful_ choice is to allow an extension type to hide the underlying representation type?

For example, what if we're creating `extension type E(Error it) ...` exactly because we want to catch certain errors, and we want to use the extension type for that purpose (to hide that we're catching an `Error`) rather than using `// ignore: avoid_catching_errors` again and again?

Perhaps we can come up with a scenario like this, but I tend to think that we should ignore it (and just rely on the representation type when the extension type is guaranteed to be forgotten anyway).

---

An interesting case is `await_only_futures`.

It is actually a [compile-time error](https://github.com/dart-lang/language/blob/main/accepted/future-releases/extension-types/feature-specification.md#:~:text=extension%20type.-,It%20is%20a%20compile%2Dtime%20error,-if%20await%20e) if we encounter `await e` where the static type `E` of `e` is an extension type that doesn't have `implements Future<T>` for any `T`.

So there's no need to decide on what to do in the lint if this situation arises.

If `E` is an extension type that _does_ have `implements ... Future<T> ...` for some `T` then `await e` will already have the desired static type, so that should just work.

On the other hand, `unawaited_futures` should probably rely on the ultimate representation type, and flag every situation where a future is left unawaited, including cases where the unawaited object has a static type which is an extension type whose ultimate representation type would match the criterion used by `unawaited_futures`.

Again, the point is that we'd want to detect situations where an unexpected run-time error occurs because a future was left unawaited (and nobody was around to receive an exception), plus situations where a program has execution ordering bugs because some operations aren't awaiting the completion of other operations. In both cases, the undesirable outcome depends on the given object being a future, not on its static type.

Again, we could come up with a scenario where we _want_ to use the extension type to hide that a certain future is a future (because we know that all these futures do not need to be awaited). However, in that case it's probably possible to avoid returning that future in the first place, or it could have the type `void` (not `Future<void>`). So I still think it makes sense to let `unawaited_futures` use the ultimate representation type when it encounters an extension type, no exceptions.

---

Otherwise, I think an extension type should probably always be preserved with respect to lints and similar checks. This supports the ability of extension types to arrange for a different treatment of the underlying object than that which is given to the representation type, and that's the main point of having extension types in the first place.

**Comment by bwilkerson** [comment]
Thanks, that's helpful.

I mentioned this elsewhere, but just to capture it somewhere more permanent: In order to reduce user confusion,  I think it's important that we document whenever a lint will use the representation type rather than the extension type.

**Comment by srawlins** [comment]
[unreachable_from_main](https://dart.dev/tools/linter-rules/unreachable_from_main) needs some changes to support extension types:

* ```dart
  extension type E(int it) {
    void m() {} // Should be reported as unreachable.
  }
  void main() { E(7); }
  ```
* ```dart
  extension type E(int it) {} // Should not be reported as unreachable.
  void main() { 7 as E; }
  ```
* ```dart
  extension type E(int it) {
    E.named(this.it); // Should be reported as unreachable.
  }
  void main() { IntExtension(7); }
  ```

**Comment by lrhn** [comment]
The description of `unreachable_from_main` says "top level or static", and *technically* `m` in the first example is neither.
That might just be because the distinction should be between statically resolved and dynamically resolved, and the "instance members" of extensions and extension types are statically resolved, even if they are not `static`.
Should the description be updated?

**Comment by srawlins** [comment]
Good catch. I think we didn't implement for instance members to avoid complexity with inheritance. I agree with your assessment and suggestion.

**Comment by srawlins** [comment]
@pq Should I add checklist items to the tippy top of this issue for the rules that I think need updating, which I commented on above?

**Comment by pq** [comment]
Thanks for bumping this, Sam. I've been meaning to! Please add items. My sense is that a bunch of the ones above are WAI as of my last conversations with @bwilkerson where I thought there was a general consensus that we did not want to lint based on runtime/representation type but I do see a few that don't fit that pattern. Thanks!

**Comment by srawlins** [comment]
OK, they're added. Some need work and some can probably just be checked because we decided above that I was wrong in the repro case I had written down.

**Comment by pq** [comment]
Ok, cool. Thanks, Sam!
